pennylane/devices/qubit/apply_operation.py

```diff
@@ -273,31 +295,50 @@ def apply_mid_measure(
     mid_measurements = execution_kwargs.get("mid_measurements", None)
     rng = execution_kwargs.get("rng", None)
     prng_key = execution_kwargs.get("prng_key", None)
-
     if is_state_batched:
         raise ValueError("MidMeasureMP cannot be applied to batched states.")
-    if not np.allclose(np.linalg.norm(state), 1.0):
-        mid_measurements[op] = -1
-        return np.zeros_like(state)
     wire = op.wires
-    sample = qml.devices.qubit.sampling.measure_with_samples(
-        [qml.sample(wires=wire)], state, Shots(1), rng=rng, prng_key=prng_key
-    )
-    sample = int(sample[0])
-    mid_measurements[op] = sample
-    if op.postselect is not None and sample != op.postselect:
-        mid_measurements[op] = -1
-        return np.zeros_like(state)
     axis = wire.toarray()[0]
     slices = [slice(None)] * qml.math.ndim(state)
-    slices[axis] = int(not sample)
-    state[tuple(slices)] = 0.0
-    state_norm = np.linalg.norm(state)
-    state = state / state_norm
-    if op.reset and sample == 1:
-        state = apply_operation(
-            qml.X(wire), state, is_state_batched=is_state_batched, debugger=debugger
-        )
+    slices[axis] = 0
+    prob0 = qml.math.cast(qml.math.norm(state[tuple(slices)]) ** 2, float)
+    interface = qml.math.get_deep_interface(state)
+    if prng_key is not None:
+        # pylint: disable=import-outside-toplevel
+        from jax.random import binomial
+
+        def binomial_fn(n, p):
+            return binomial(prng_key, n, p).astype(int)
+
+    else:
+        binomial_fn = np.random.binomial if rng is None else rng.binomial
+    sample = binomial_fn(1, 1 - prob0)
+    mid_measurements[op] = sample
+
+    # Using apply_operation(qml.QubitUnitary,...) instead of apply_operation(qml.Projector([sample], wire),...)
+    # to select the sample branch enables jax.jit and prevents it from using Python callbacks
+    matrix = qml.math.array([[(sample + 1) % 2, 0.0], [0.0, (sample) % 2]], like=interface)
```

Comment:
This is not a unitary, right? But we're able to get away with this because there's no validation of the matrix?

Reply:
Another merge artifact. This is the matrix for a projector to apply to the state for postselection.